### PR TITLE
Maximo/measured metric

### PR DESCRIPTION
**Pull request by link04**
What does this PR do?
---------------------
Adds automatic logic to tag spans for metrics calculation by adding "_dd.measure" and also when users tag their spans manually if they need metric calculation for their custom spans.

Who will it impact?
-------------------
All tracer users and backend Agent team

Motivation
----------
Internal Tracking: 
https://datadoghq.atlassian.net/browse/DOTNET-77 
https://docs.google.com/spreadsheets/d/1EXrW-CnPqg-JPlIPMB4P055xwovhzeKFUGnEh-lZcTQ/edit#gid=817210442

Test
-----
Tested locally for the following integrations:
- WCF (server) - wcf.request - https://a.cl.ly/YEuQ8rAE 
- WebClient/Web.Request - http.request - https://a.cl.ly/JruqXOLQ  
- HttpClient/HttpHandler - http.request - https://a.cl.ly/4guO5KJ7 
- GraphQL (server) - graphql.validate - https://a.cl.ly/d5uPz9pk 
- GraphQL (server) - graphql.execute - https://a.cl.ly/2Nu02z0r 

@DataDog/apm-dotnet

### CODE REVIEW

**Review by kevingosse — CHANGES_REQUESTED** [review]
It's unclear to me what rationale was use to decorate integrations with this tag or not. For instance, if GraphQL is decorated, I don't understand why ElasticSearch isn't?

**Comment by lucaspimentel** [comment]
closing and re-opening to try to fix the diff.

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/GraphQLIntegration.cs [hunk 1]
```diff
@@ -219,6 +219,7 @@ public static object ExecuteAsync(object executionStrategy, object context, int
         private static void DecorateSpan(Span span, GraphQLTags tags)
         {
             span.Type = SpanTypes.GraphQL;
+            span.SetMetric(Tags.Measured, 1);
```
[the remark this answers is not in the document]
Comment (link04) [reply src/Datadog.Trace.ClrProfiler.Managed/Integrations/GraphQLIntegration.cs]: Thank you for the note Kevin, I've addressed this in in commit **370a111**.

**Comment by lucaspimentel** [comment]
> It's unclear to me what rationale was used to decorate integrations with this tag or not.

@kevingosse: After some confusion, we determined that we should add the new metric to a span if it is either:
- a root span (i.e. web framework spans)
- a span that currently has a different service name than the root span but will soon be changed to have the same name as the root span (i.e. `myservice-sql-server` will be renamed to `myservice`)

By this logic, all our current integrations should have this metric. However, in the future we will have integrations that are neither root spans nor external calls, and those should not have this metric. (@brettlangdon, please correct me if I'm wrong.)

**Review by kevingosse — APPROVED** [review]
LGTM

File: src/Datadog.Trace.ClrProfiler.Managed/Integrations/WcfIntegration.cs [hunk 2]
```diff
@@ -163,6 +163,7 @@ private static Scope CreateScope(object requestContext)
                     tags,
                     tagsFromHeaders);
 
+                tags.Measured = 1;
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Integrations/WcfIntegration.cs:166]: Looks like we already add the metric in `span.DecorateWebServerSpan(...)` above, so we don't need to add it again here.

File: src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs [hunk 3]
```diff
@@ -289,6 +289,7 @@ private void OnHostingHttpRequestInStart(object arg)
 
             scope.Span.DecorateWebServerSpan(resourceName, httpMethod, host, url, tags, tagsFromHeaders);
 
+            tags.Measured = 1;
```
Comment (lucaspimentel) [inline src/Datadog.Trace/DiagnosticListeners/AspNetCoreDiagnosticObserver.cs:292]: Looks like we already add the metric in `span.DecorateWebServerSpan(...)` above, so we don't need to add it again here.

**Review by lucaspimentel — CHANGES_REQUESTED** [review]
Looking good so far! There's a few integrations that are setting the new metric twice (see comments). We should consolidate those to the helper method so it's only added in one place.

**Comment by link04** [comment]
Created new PR to address the objective of only adding the option for the customers to set the metric manually since it will not be added to any integrations by automatically: https://github.com/DataDog/dd-trace-dotnet/pull/1066

**Pull request closed without merging**